src/main/kotlin/ecommerce/service/ProductService.kt

```diff
@@ -0,0 +1,78 @@
+package ecommerce.service
+
+import ecommerce.dto.OptionDto
+import ecommerce.dto.ProductPatchRequest
+import ecommerce.dto.ProductRequest
+import ecommerce.dto.ProductResponse
+import ecommerce.dto.toEntity
+import ecommerce.exception.ProductCreationException
+import ecommerce.exception.ProductNotFoundException
+import ecommerce.exception.ProductUpdateException
+import ecommerce.model.toDto
+import ecommerce.repository.ProductJpaRepository
+import ecommerce.repository.existsByNameOrThrow
+import ecommerce.repository.getByIdOrThrow
+import org.springframework.data.domain.Page
+import org.springframework.data.domain.Pageable
+import org.springframework.stereotype.Service
+
+@Service
+class ProductService(private val productJpaRepository: ProductJpaRepository) {
+    fun findById(id: Long): ProductResponse {
+        val product = productJpaRepository.getByIdOrThrow(id)
+        return product.toDto()
+    }
+
+    fun findAll(pageable: Pageable): Page<ProductResponse> {
+        val products = productJpaRepository.findAll(pageable)
+
+        return products.map { product -> product.toDto() }
+    }
+
+    fun createProduct(productRequest: ProductRequest): ProductResponse {
+        productJpaRepository.existsByNameOrThrow(productRequest.name)
+        try {
+            return productJpaRepository.save(productRequest.toEntity()).toDto()
+        } catch (e: Exception) {
+            throw ProductCreationException("Failed to create product")
+        }
+    }
+
+    fun updateProduct(
+        id: Long,
+        productRequest: ProductPatchRequest,
+    ): ProductResponse {
+        val product = productJpaRepository.getByIdOrThrow(id)
+        val newProduct =
+            product.copy(
+                name = productRequest.name ?: product.name,
+                price = productRequest.price ?: product.price,
+                imageUrl = productRequest.imageUrl ?: product.imageUrl,
+                options = product.options,
+            )
```

Comment:
```suggestion
            product.copy(
                name = productRequest.name ?: product.name,
                price = productRequest.price ?: product.price,
                imageUrl = productRequest.imageUrl ?: product.imageUrl,
            )
```

How does `copy()` work?
To me, putting `product.options` here seems redundant 🙂

Reply:
You are right!
Because copy() copies all properties by default from the original object unless I explicitly override them, what makes the property "options" redundant and unnecessary.